[CMake] Create tool header collector target only when building tools

kateinoigakukun opened the pull request:
`HeaderDependencies.cpp` cannot be compiled with `SWIFT_INCLUDE_TOOLS=OFF` because it depends on `Config.h` which is only generated while building tools. Exclude it from the build to fix `all` target.

Follow-up to b39dba32f39bd6f0987beed98465ff55de6b3d19

Hopefully this fixes https://ci.swift.org/job/oss-swift-test-stdlib-with-toolchain/7124

```
In file included from /Users/ec2-user/jenkins/workspace/oss-swift-test-stdlib-with-toolchain/build/stdlib_RA_standalone/swift-macosx-x86_64/include/HeaderDependencies.cpp:6:
In file included from /Users/ec2-user/jenkins/workspace/oss-swift-test-stdlib-with-toolchain/swift/include/swift/AST/ASTBridging.h:26:
In file included from /Users/ec2-user/jenkins/workspace/oss-swift-test-stdlib-with-toolchain/swift/include/swift/AST/DiagnosticEngine.h:21:
In file included from /Users/ec2-user/jenkins/workspace/oss-swift-test-stdlib-with-toolchain/swift/include/swift/AST/ActorIsolation.h:19:
In file included from /Users/ec2-user/jenkins/workspace/oss-swift-test-stdlib-with-toolchain/swift/include/swift/AST/Type.h:23:
In file included from /Users/ec2-user/jenkins/workspace/oss-swift-test-stdlib-with-toolchain/swift/include/swift/AST/LayoutConstraint.h:22:
In file included from /Users/ec2-user/jenkins/workspace/oss-swift-test-stdlib-with-toolchain/swift/include/swift/AST/PrintOptions.h:17:
/Users/ec2-user/jenkins/workspace/oss-swift-test-stdlib-with-toolchain/swift/include/swift/AST/AttrKind.h:22:10: fatal error: 'swift/Config.h' file not found
#include "swift/Config.h"
         ^~~~~~~~~~~~~~~~
1 error generated.
```

Resolves rdar://120907760

kateinoigakukun committed:
[CMake] Create tool header collector target only when building tools

`HeaderDependencies.cpp` cannot be compiled with
`SWIFT_INCLUDE_TOOLS=OFF` because it depends on `Config.h` which is only
generated while building tools. Exclude it from the build to fix `all`
target.

Follow-up to b39dba32f39bd6f0987beed98465ff55de6b3d19

kateinoigakukun commented:
@swift-ci Please test stdlib with toolchain

kateinoigakukun commented:
Hmm, it seems like we need to update Jenkins pipeline configuration of `swift-PR-stdlib-with-toolchain-osx` to include `cmake` in `PATH` as other pipelines do. CC @shahmishal 

```
+ /Users/ec2-user/jenkins/workspace/swift-PR-stdlib-with-toolchain-osx/branch-main/swift/utils/build-script --preset=stdlib_RA_standalone,build,test toolchain_path=/Users/ec2-user/jenkins/workspace/swift-PR-stdlib-with-toolchain-osx/branch-main/latest_toolchain/usr/bin/
[/Users/ec2-user/jenkins/workspace/swift-PR-stdlib-with-toolchain-osx/branch-main/swift/utils/build-script] NOTE: using preset "stdlib_RA_standalone,build,test", which expands to 

/Users/ec2-user/jenkins/workspace/swift-PR-stdlib-with-toolchain-osx/branch-main/swift/utils/build-script --skip-build-llvm --skip-build-cmark --skip-build-benchmarks --skip-test-cmark --build-swift-tools=0 --build-swift-libexec=0 --skip-early-swift-driver --skip-early-swiftsyntax --native-swift-tools-path=/Users/ec2-user/jenkins/workspace/swift-PR-stdlib-with-toolchain-osx/branch-main/latest_toolchain/usr/bin/ --native-llvm-tools-path=/Users/ec2-user/jenkins/workspace/swift-PR-stdlib-with-toolchain-osx/branch-main/latest_toolchain/usr/bin/ --native-clang-tools-path=/Users/ec2-user/jenkins/workspace/swift-PR-stdlib-with-toolchain-osx/branch-main/latest_toolchain/usr/bin/ --build-ninja '--lit-args=--filter='"'"':: (stdlib/|Concurrency/)'"'"' -v' --only-executable-test --build-subdir=stdlib_RA_standalone --release --assertions --verbose-build --test --validation-test

[/Users/ec2-user/jenkins/workspace/swift-PR-stdlib-with-toolchain-osx/branch-main/swift/utils/build-script] NOTE: Using toolchain default
+ /usr/libexec/PlistBuddy -c 'Print :SupportedTargets:macosx:Archs' /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/SDKSettings.plist
[/Users/ec2-user/jenkins/workspace/swift-PR-stdlib-with-toolchain-osx/branch-main/swift/utils/build-script] ERROR: can't find CMake (please install CMake)
```
https://ci.swift.org/job/swift-PR-stdlib-with-toolchain-osx/8/console

rintaro commented:
Huh, sorry I didn't notice. Maybe we should just this to `lib/CMakeLists.txt`

kateinoigakukun commented:
Okay, putting it in `lib/CMakeLists.tx` sounds reasonable to me 👍

kateinoigakukun commented:
@swift-ci Please smoke test

rintaro commented on lib/CMakeLists.txt:
```diff
@@ -76,6 +76,34 @@ if (SWIFT_BUILD_SWIFT_SYNTAX)
+#include \"swift/SILOptimizer/OptimizerBridging.h\"
+")
+add_library(importedHeaderDependencies "${CMAKE_CURRENT_BINARY_DIR}/HeaderDependencies.cpp")
+
+# When building unified, we need to make sure all the Clang headers are
+# generated before we try to use them.
+# When building Swift against an already compiled LLVM/Clang, like
+# build-script does, this target does not exist, but the headers
+# are already completely generated at that point.
+if(TARGET clang-tablegen-targets)
+  add_dependencies(importedHeaderDependencies clang-tablegen-targets)
+endif()
```
```suggestion
add_dependencies(importedHeaderDependencies LLVM_COMMON_DEPENDS)
```
I realized `clang-tablegen-targets` is always available. and I _think_ `LLVM_COMMON_DEPENDS` would be better here. 
https://github.com/apple/swift/blob/c8fe36d226cf7f2eb7913e10f9029ed0b8cfa46b/lib/CMakeLists.txt#L1-L6

kateinoigakukun replied:
That makes sense *to me*

kateinoigakukun committed:
[CMake] Cleanup TableGen dependency handling

Clang and LLVM always provide those tablegen targets, so we don't need
to check their existence. Also adding `intrinsic_gen` as a dependency.
See https://github.com/apple/swift/pull/12147

kateinoigakukun commented:
@swift-ci Please smoke test

tshortli commented:
Thanks for the fix! I'm enabling auto-merge since this fixes some CI failures and I'd like it to get in quickly.
